src/Popova/Task5/main.cpp

```diff
@@ -0,0 +1,179 @@
+#include <QVector3D>
+#include <cmath>
+#include "Types.hpp"
+#include "Sphere.hpp"
+#include "Plane.hpp"
+#include <QImage>
+#include <QColor>
+#include <QOpenGLFunctions>
+#include "QCoreApplication"
+#include <QCommandLineParser>
+#include <iostream>
+
+int WIDTH = 640;
+int HEIGHT = 480;
+constexpr float PI = 3.141592F;
+constexpr auto MAX_REFLECTIONS = 4;
+constexpr QVector3D BACKGROUND_COLOR{0.7f, 0.7f, 0.9f};
+constexpr QVector3D LIGHT_COLOR{1.0F, 1.0F, 1.0F};
+constexpr QVector3D LIGHT_POS{0.0F, 4.0F, 0.0F};
+constexpr float LIGHT_INTENSITY = 1.5F;
+constexpr QVector3D CAMERA_POS{0.0F, 1.0F, 3.0F};
+
+QVector3D mix(const QVector3D &a, const QVector3D &b, const float k){
+    return a * k + b*(1 - k);
+}
+
+
+QVector3D reflect(const QVector3D &incident, const QVector3D &normal) {
+return incident - 2.f * normal * (QVector3D::dotProduct(incident, normal));
+}
+
+
+QVector3D refract(const QVector3D &incident, const QVector3D &normal, const float n_1, const float n_2=1.f) {
+    auto cos_1 = - std::max(-1.f, std::min(1.f, QVector3D::dotProduct(incident, normal)));
+    if (cos_1 < 0.f) return refract(incident, -normal, n_2, n_1);
+    auto ratio = n_2 / n_1;
+    auto cos_2 = sqrtf(1 - ratio * ratio * (1 - cos_1 * cos_1));
+    return  ratio * incident + (ratio * cos_1 - cos_2) * normal;
+}
+
+
+bool intersect(const Ray &ray, const std::vector<Sphere> &spheres, const std::vector<Plane> &planes, QVector3D &hit, QVector3D &normal, Material &material, const QImage &texture) {
+    auto spheres_dist = std::numeric_limits<float>::max();
+    // QImage TEXTURE = QImage(":/textures/texture.jpg");
+    for (const auto & sphere : spheres) {
+        auto dist_i = std::numeric_limits<float>::max();
+        if (sphere.intersect(ray, dist_i) && dist_i < spheres_dist) {
+            spheres_dist = dist_i;
+            hit = ray.origin + ray.direction * dist_i;
+            normal = (hit - sphere.center).normalized();
+            material = sphere.material;
+           
+        }
+    }
+    float planes_dist = std::numeric_limits<float>::max();
+    for (const auto & plane : planes) {
+        auto dist_i = std::numeric_limits<float>::max();
+        if (plane.intersect(ray, dist_i) && dist_i < spheres_dist) {
+            planes_dist = dist_i;
+            hit = ray.origin + ray.direction * dist_i;
+            normal = plane.normal;
+            int width = texture.width()-1;
```

Comment:
```suggestion
            const auto width = texture.width()-1;
```
и в остальных местах по коду